Manual: built-in attributes and extension node; extension literals

Octachron opened the pull request:
This pull request adds a brief description for the new built-in attributes and extension node. The covered built-in attributes are
-`[@tailcall]`
-`[@@inline]`
-`[@inlined]`
-`[@warn_on_literal_pattern]`
The built-in attributes related to the c interface (i.e. `[@unboxed]`, `[@untaged]` and `[@@noalloc]`) are also mentioned in the built-in attributes subsection and the reader is referred to the related subsection for more information.

The description of the built-in extension constructor `[%extension_constructor]` has been added to a new subsection in the extension nodes section.

Moreover, the new literal modifiers for int and float literals (e.g. `1m`) are described in a new extension literal subsection. This subsection has been grouped with the extension operators subsection to form an extension-only syntax section.

Octachron committed:
Manual: more built-in attributes

Add a description for
-inline
-inlined
-tailcall
-warn_on_literal_pattern

and a cross-reference for
-noalloc
-unboxed
-untagged

Octachron committed:
Manual: built-in [%extension_constructor]

Octachron committed:
Manual: extension-only syntax, extension literals

Octachron committed:
Manual: add a line break to hex float literals

Octachron committed:
Manual: better tailcall description

Octachron commented on manual/manual/refman/exten.etex:
```diff
@@ -1539,6 +1539,28 @@ Some attributes are understood by the type-checker:
+\item
+  ``ocaml.tailcall'' or ``tailcall'' can be applied to function
+  application in order to indicate that the call should be
+  tailcall optimized. If it it not the case, the warning 51 is emitted.
```
You are right. Fixed.

Octachron commented on manual/manual/refman/exten.etex:
```diff
@@ -1546,16 +1568,31 @@ module X = struct
+
+let rec no_op = function
+  | [] -> []
+  | _ :: q -> (no_op[@tailcall]) q;;
```
Basic typographic error,  `[ ]` was supposed to be `( )`.

Octachron committed:
Manual: fix typo in code examples

in built-in attributes and extension subsection

Octachron commented on manual/manual/refman/exten.etex:
```diff
@@ -1641,6 +1678,27 @@ the attributes are considered to apply to the payload:
+type t += X of int | Y of string
+let x = [%extension_constructor X]
+let y = [%extension_constructor Y]
+assert (x <> y)
```
Good point.

gasche commented on manual/manual/refman/exten.etex:
```diff
@@ -1539,6 +1539,28 @@ Some attributes are understood by the type-checker:
+\item
+  ``ocaml.warn_on_literal_pattern'' or ``warn_on_literal_pattern'' annotate
+  constructors in type definition. The warning 53 is then emitted when this
+  constructor is pattern matched with a constant literal as argument.
```
For other attributes documented in this list, the interest of using them is obvious. For this one you should give a justification -- fragility of `failwith` arguments, etc.

Octachron replied:
I have added more details and a note to signal that the warning is triggered only for single argument
constructor.

gasche replied:
This single argument thing is a pain point. We should maybe note in the documentation that this is purely an implementation limitation, and more importantly we should raise a warning when the attribute is placed on a constructor declaration with several arguments.

Octachron replied:
Documenting the fact that this is an implementation limitation is fine with me. Raising a warning or
extending the current implementation seems like an objective for another pull request.

gasche replied:
I raised the topic in #254 .

gasche commented on manual/manual/refman/exten.etex:
```diff
@@ -1539,6 +1539,28 @@ Some attributes are understood by the type-checker:
+  If it it not the case, a warning 51 is emitted.
+\item
+  ``ocaml.inline'' or ``inline'' take either ``never'', ``always''
+  or nothing as payload on a function or functor definition. No payload
```
"No payload is equivalent to _always_" is confusing grammar. You should rather say: "If no payload is provided, the default value is _always_".

Octachron replied:
Fixed, thanks.

gasche commented on manual/manual/refman/exten.etex:
```diff
@@ -1546,16 +1568,31 @@ module X = struct
+  | String of string [@warn_on_literal_pattern]
+
+let f = function
+| Int 0 | String "constant" -> () (* trigger warning *)
```
triggers warning 53

Octachron replied:
Warning 52 in fact. I have added the warning number, and fixed it in the attribute description.

Octachron committed:
Manual: fix built-in attributes section

-more homogeneous warning notation
-fix warn_on_literal_pattern warning number
-more in-details explanation of warn_on_literal_pattern
-fix grammar for inline attribute

Octachron committed:
Manual: detail warn_on_literal_pattern limitation

gasche commented:
You ( @Octachron ) have been doing tremendous work on the documentation front, which are not appropriately credited in the Changes file. Could you add a new Changes item in the "Manual" section, mentioning that you documented the new 4.03 features, referencing the present PR# and [MPR#7092](http://caml.inria.fr/mantis/view.php?id=7092)?

Octachron commented:
I am fine with adding a line in the Changes files. 

However, crediting myself for [MPR#7092](http://caml.inria.fr/mantis/view.php?id=7092) is going too far since at least @xavierleroy and @alainfrisch wrote some of the items explicitly mentioned in this ticket. Moreover, at least, @Drup, @diml and @garrigue  also contributed to the general 4.03.0 documentation and should have the right to be mentioned.

gasche commented:
If we duplicate in the "Manual" section any change to the base language, there is going to be pushback against a Changes section too verbose (we must balance crediting and the ability for end-users to read the changelog for each new release). In the future, now that the manual is merged, we can require people that send a PR to change the language to also update the manual, and thus no separate Manual crediting is required. I therefore suggest that we credit for MPR#7092 all the people that have contributed documentation for a feature that they have not implemented themselves, and to my knowledge you are the only one in this situation.

I understand that handling Credit in this way may be uncomfortable for you, so I'll merge the PR and do the Changes part myself.
